File: src/main/java/refactown/cleancode/c01cleancode/Refactoring2.java
package refactown.cleancode.c01cleancode;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.math.BigInteger;
import java.util.*;

import static java.lang.reflect.Modifier.isFinal;
import static java.lang.reflect.Modifier.isStatic;

public class Refactoring2 {

    private static final Set<Class<?>> INTEGER_TYPES = new HashSet<>();

    static {
        INTEGER_TYPES.add(Integer.class);	INTEGER_TYPES.add(int.class);
        INTEGER_TYPES.add(Long.class);		INTEGER_TYPES.add(long.class);
        INTEGER_TYPES.add(Short.class);		INTEGER_TYPES.add(short.class);
        INTEGER_TYPES.add(Byte.class);		INTEGER_TYPES.add(byte.class);
        INTEGER_TYPES.add(BigInteger.class);
    }

    public static int countIntegerConstants(Class<?> clazz) { // +1
        if (!isClasseComConst(clazz)) { // +1
            return -1; // +1
        }

        List<Field> fields = getFieldsFromClass(clazz);

        int count = 0;
        for (Field field : fields) { // +1
            if (isIntConst(field)) { // +1
                System.out.println("Int Constant:" + field);
                count++;
            }
        }
        return count; // CC = 5 (Total = 22)
    }

    public static boolean isClasseComConst(Class<?> clazz) { // +1
        return clazz != null && !clazz.isEnum() // +1
                && !Modifier.isAbstract(clazz.getModifiers()); // +1 cc = 3
    }

    public static List<Field> getFieldsFromClass(Class<?> clazz) { // +1
        List<Field> fields = new ArrayList<>();
        fields.addAll(Arrays.asList(clazz.getDeclaredFields()));
        Class<?> superClass = clazz.getSuperclass();
        if (superClass != null && !superClass.equals(Object.class)) { // +2
            fields.addAll(getFieldsFromClass(superClass));
        }
        return fields; // CC = 3

    }

    public static boolean isIntConst(Field field) { // +1
        return INTEGER_TYPES.contains(field.getType())
                && isStatic(field.getModifiers())
                && isFinal(field.getModifiers()); // +1 CC = 3
    }


}

File: src/main/kotlin/refactown/cleancode/c01cleancode/Refactoring2.kt
package refactown.cleancode.c01cleancode

import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.math.BigInteger

fun countIntegerConstants(clazz: Class<*>): Int {
    if (!isConstantClass(clazz)) return -1
    val fields = getFieldsFromClass(clazz)
    var count = 0
    for (field in fields) {
        if (isIntegerField(field)) {
            println("Int Constant:$field"); count++
        }
    }
    return count
}

fun isConstantClass(c: Class<*>?) = c != null && !c.isEnum && !Modifier.isAbstract(c.modifiers)

fun getFieldsFromClass(clazz: Class<*>): List<Field> {
    return if (clazz.superclass != null && clazz.superclass != Any::class.java) {
        clazz.declaredFields.asList() + getFieldsFromClass(clazz.superclass)
    } else clazz.declaredFields.asList()
}

val integerTypes = listOf(Int::class.java, Int::class.javaPrimitiveType,
        Long::class.java, Long::class.javaPrimitiveType, Short::class.java,
        Short::class.javaPrimitiveType, Byte::class.java, Byte::class.javaPrimitiveType,
        BigInteger::class.java)

fun isIntegerField(field: Field): Boolean = integerTypes.contains(field.type)
        || Modifier.isStatic(field.modifiers) || Modifier.isFinal(field.modifiers)



